### PR TITLE
Add advanced bifacial PV and ground irradiance modeling

### PR DESCRIPTION
**Pull request by Matthew-Boyd**
Adds:
- Irradiance loss factor differentiation
- Rack shading loss
- Intra-module electrical mismatch
- Multi-albedo ground
- Temporospatial ground and rear irradiance output
- Bifacial loss diagram additions

### CODE REVIEW

**Review by mjprilliman — COMMENTED** [review]
I haven't come across any issues.

File: shared/lib_irradproc.h [hunk 1]
```diff
@@ -867,6 +868,32 @@ void ModifiedDISC(const double kt[3], const double kt1[3], const double g[3], co
...
 
+/**
+* divideAndAlignAlbedos subdivides the spatial albedo vector and if 1-axis tracking
+* changes reference from the row midline to the front
+*
+* \param[in] albedo spatial albedo for that point in time
+* \param[in] n_divisions increased number of divisions to divide the spatial albedo into
+* \param[in] isOneAxisTracking is 1-axis tracking used?
+* \param[in] horizontalLength projected horizontal length of the row, product of slope length and tilt
+* \param[in] rowToRow distance between front of row and front of row behind
```
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply shared/lib_irradproc.h]: Done.

File: shared/lib_irradproc.cpp [hunk 2]
```diff
@@ -3128,6 +3255,82 @@ double shadeFraction1x(double solar_azimuth, double solar_zenith,
     return fs;
 }
 
+std::vector<double> divideAndAlignAlbedos(const std::vector<double>& albedo /*-*/, size_t n_divisions /*-*/, bool isOneAxisTracking /*-*/,
```
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply shared/lib_irradproc.cpp:3258]: No, they were created for this feature.

File: shared/lib_pv_io_manager.cpp [hunk 3]
```diff
@@ -1329,7 +1358,16 @@ Module_IO::Module_IO(compute_module* cm, std::string cmName, double dcLoss)
 
     }
     else
+    {
         throw exec_error(cmName, "invalid pv module model type");
+    }
+
+    // TODO: reimplement, but fix issues that this causes with some tests
```
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply shared/lib_pv_io_manager.cpp:1365]: The commented out code I initially added should of course not cause an issue, but it does. I'm just pointing it out here, but I wasn't able to quickly troubleshoot why.

File: ssc/cmod_pvsamv1.cpp [hunk 4]
```diff
@@ -80,33 +86,41 @@ static var_info _cm_vtab_pvsamv1[] = {
...
         {SSC_INPUT, SSC_NUMBER,   "subarray2_nameplate_loss",             "Sub-array 2 DC nameplate loss",                       "%",      "",                                                                                                                                                                                      "Losses",                                                "?",                                  "MIN=-5,MAX=100",      "" },
+        {SSC_INPUT, SSC_NUMBER,   "subarray2_electrical_mismatch",        "Sub-array 2 electrical mismatch loss",                "%",      "",                                                                                                                                                                                      "Losses",                                                "?",                                  "MIN=0,MAX=100",       "" },
```
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply ssc/cmod_pvsamv1.cpp]: They could but aren't to keep the variable name short. I added 'bifacial' to the labels.

**Review by mjprilliman — APPROVED** [review]

**Pull request merged**